File: src/platform/ios/sem.m
//
//  sem.c
//  baye-ios
//
//  Created by loong on 15/8/15.
//
//

#import <Foundation/Foundation.h>

#include "sem.h"

GAM_SEM gam_sem_create()
{
    return (__bridge_retained void*)dispatch_semaphore_create(0);
}

void gam_sem_delete(GAM_SEM sem)
{
    (void)(__bridge_transfer dispatch_semaphore_t)sem;
}

void gam_sem_signal(GAM_SEM sem)
{
    dispatch_semaphore_signal((__bridge dispatch_semaphore_t)sem);
}

void gam_sem_wait(GAM_SEM sem)
{
    dispatch_semaphore_wait((__bridge dispatch_semaphore_t)sem, DISPATCH_TIME_FOREVER);
}

GAM_LOCK gam_lock_create() {
    GAM_LOCK lck = gam_sem_create();
    gam_lock_unlock(lck);
    return lck;
}
void gam_lock_delete(GAM_LOCK lck) { gam_sem_delete(lck); }
void gam_lock_lock(GAM_LOCK lck) { gam_sem_wait(lck); }
void gam_lock_unlock(GAM_LOCK lck) { gam_sem_signal(lck); }

File: src/platform/js/sem.c
//
//  sem.c
//  baye-ios
//
//  Created by loong on 15/8/15.
//
//


#include "sem.h"
#include <stdlib.h>
#include <emscripten.h>
#include <baye/compa.h>

GAM_SEM gam_sem_create()
{
    int* p = (int*)gam_malloc(sizeof(int));
    *p = 0;
    return (GAM_SEM)p;
}

void gam_sem_delete(GAM_SEM sem)
{
    gam_free((void*)sem);
}

void gam_sem_signal(GAM_SEM sem)
{
    *((int*)sem) += 1;
}

void gam_sem_wait(GAM_SEM sem)
{
    int *p = (int*)sem;
    emscripten_sleep_with_yield(1);
    while (*p == 0) {
        emscripten_sleep_with_yield(10);
    }
    *p -= 1;
}


GAM_LOCK gam_lock_create() {
    return NULL;
}
void gam_lock_delete(GAM_LOCK lck) {}
void gam_lock_lock(GAM_LOCK lck) {}
void gam_lock_unlock(GAM_LOCK lck) {}
